src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -195,7 +202,7 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
         private void CaptureVariable(Symbol variable, SyntaxNode syntax)
         {
             var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
-            if (type.IsRestrictedType())
+            if (type.IsRestrictedType() || variable is LocalSymbol { RefKind: not RefKind.None, IsCompilerGenerated: false })
```

Comment:
> || variable is LocalSymbol { RefKind: not RefKind.None, IsCompilerGenerated: false } [](http://example.com/codeflow?start=39&length=85)

Why do we need to make any changes here? I thought MethodToClassRewriter already reports unsupported spilling

Reply:
> I thought MethodToClassRewriter already reports unsupported spilling

`MethodToStateMachineRewriter` reports `ERR_RefReturningCallAndAwait` but that's only for calls.

Otherwise `MethodToClassRewriter` and similar rely on list of hoisted variables created by this `IteratorAndAsyncCaptureWalker` - so this change makes sure ref locals are not hoisted. Previously that was not necessary as ref locals were disallowed during binding in async methods.